src/wifi_dpp.rs

```diff
@@ -0,0 +1,223 @@
+                warn!("Internal error receiving event!");
+                EspError::from_infallible::<ESP_ERR_INVALID_STATE>()
+            })?;
+        debug!("dpp_bootstrap_gen got: {event:?}");
+        match event {
+            DppEvent::UriReady(qrcode) => {
+                // Bit of a hack to put the wifi driver in the correct mode.
+                self.ensure_config_and_start()?;
+                Ok(EspDppBootstrapped::<QrCode> {
+                    events_rx: &self.events_rx,
+                    data: QrCode(qrcode),
+                })
+            }
+            _ => {
+                warn!("Got unexpected event: {event:?}");
+                Err(EspError::from_infallible::<ESP_ERR_INVALID_STATE>())
+            },
+        }
+    }
+
+    fn ensure_config_and_start(&mut self) -> Result<ClientConfiguration, EspError> {
+        let operating_config = match self.wifi.get_configuration()? {
+            Configuration::Client(c) => c,
+            _ => {
+                let fallback_config = ClientConfiguration::default();
+                self.wifi.set_configuration(&Configuration::Client(fallback_config.clone()))?;
+                fallback_config
+            },
+        };
+        if !self.wifi.is_started()? {
+            self.wifi.start()?;
+        }
+        Ok(operating_config)
+    }
+
+    unsafe extern "C" fn dpp_event_cb_unsafe(
+        evt: esp_supp_dpp_event_t,
+        data: *mut ::core::ffi::c_void
+    ) {
+        debug!("dpp_event_cb_unsafe: evt={evt}");
+        let event = match evt {
+            esp_supp_dpp_event_t_ESP_SUPP_DPP_URI_READY => {
+                DppEvent::UriReady(CStr::from_ptr(data as *mut c_char).to_str().unwrap().into())
+            },
+            esp_supp_dpp_event_t_ESP_SUPP_DPP_CFG_RECVD => {
+                let config = data as *mut wifi_config_t;
+                // TODO: We're losing pmf_cfg.required=true setting due to missing
+                // information in ClientConfiguration.
+                DppEvent::ConfigurationReceived(Newtype((*config).sta).into())
+            },
+            esp_supp_dpp_event_t_ESP_SUPP_DPP_FAIL => {
+                DppEvent::Fail(EspError::from(data as esp_err_t).unwrap())
+            }
+            _ => panic!(),
+        };
+        dpp_event_cb(event)
+    }
+}
+
+fn dpp_event_cb(event: DppEvent) {
+    match EVENTS_TX.lock().deref() {
+        Some(tx) => {
+            debug!("Sending: {event:?}");
+            if let Err(e) = tx.try_send(event) {
+                error!("Cannot relay event: {e}");
+            }
+        }
+        None => warn!("Got spurious {event:?} ???"),
+    }
+}
+
+
+#[derive(Debug)]
+enum DppEvent {
+    UriReady(String),
+    ConfigurationReceived(ClientConfiguration),
+    Fail(EspError),
+}
+
+impl<'d, 'w> Drop for EspDppBootstrapper<'d, 'w> {
+    fn drop(&mut self) {
+        unsafe { esp_supp_dpp_deinit() };
+    }
+}
+
+pub struct EspDppBootstrapped<'d, T> {
+    events_rx: &'d Receiver<DppEvent>,
```

Comment:
Hmm, this should probably be &mut so that it guarantees the bootstrapper binding cannot be used while the bootstrapped binding is in scope.